src/Data/Generic/HKD/Named.hs

```diff
@@ -0,0 +1,106 @@
+{-# OPTIONS_HADDOCK not-home #-}
+
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-|
+Module      : Data.Generic.HKD.Named
+Description : Construct an HKD record with named parameters.
+Copyright   : (c) Tom Harding, 2019
+License     : MIT
+Maintainer  : tom.harding@habito.com
+Stability   : experimental
+-}
+module Data.Generic.HKD.Named
+  ( Record (..)
+  ) where
+
+import Data.Functor.Contravariant (Contravariant (..))
+import Data.Generic.HKD.Types (HKD, HKD_)
+import Data.GenericLens.Internal (GUpcast (..))
+import Data.Kind (Type)
+import GHC.Generics
+import Named ((:!), NamedF (..))
+
+class Append (xs :: Type -> Type) (ys :: Type -> Type) (zs :: Type -> Type)
+  | xs ys -> zs where
+
+instance Append (S1 m h) t (S1 m h :*: t)
+instance Append x y z => Append (w :*: x) y (w :*: z)
+
+class Rearrange (i :: Type -> Type) (o :: Type -> Type) | i -> o
+instance Rearrange i o => Rearrange (D1 m i) (D1 m o)
+instance Rearrange i o => Rearrange (C1 m i) (C1 m o)
+instance Rearrange i o => Rearrange (S1 m i) (S1 m o)
+instance                  Rearrange (Rec0 x) (Rec0 x)
+instance (Rearrange l l', Rearrange r r', Append l' r' o)
+    => Rearrange (l :*: r) o where
+
+-- | The 'Data.Generic.HKD.build' function lets us supply arguments to a type
+-- one by one, but can cause confusion when working with a record. If the
+-- record contains two fields of the same type, for example, we've introduced
+-- an opportunity for bugs and confusion. The @record@ function uses the
+-- wonderful @named@ package to help us:
+--
+-- >>> :set -XDeriveGeneric -XTypeApplications
+--
+-- >>> :{
+-- data User
+--   = User { name :: String, age :: Int, likesDogs :: Bool }
```

Comment:
Shouldn't the example use fields of the same type but with different names to follow the motivation above?